### PR TITLE
[POAE7-2582] support loop join for arrow format

### PR DESCRIPTION
**Pull request by spevenhe**
### What changes were proposed in this pull request?

1. support loop join for arrow format
2. updated loop join codegen for left join type
3. commented out columnvar fetch
### Why are the changes needed?

support arrow format

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Uts

### CODE REVIEW

File: cider/tests/functionality/CiderNonEqualLoopJoinTest.cpp [hunk 1]
```diff
@@ -22,164 +22,215 @@
 #include <gtest/gtest.h>
 #include "tests/utils/CiderTestBase.h"
 
-class CiderNonEqualJoinTest : public CiderJoinTestBase {
+// TODO: (spevenhe) For not null query, if set NOT NULL on column ddl, the final project
+// will go to cider_agg_id_proj_xx() instead of cider_agg_id_proj_xx_nullable(), So the
+// returned null values are incorrect.
```
Comment (jikunshang) [inline cider/tests/functionality/CiderNonEqualLoopJoinTest.cpp:27]: why results are incorrect?

File: cider/tests/functionality/CiderNonEqualLoopJoinTest.cpp [hunk 2]
```diff
@@ -22,164 +22,215 @@
...
+class CiderArrowSeqNotNullJoinTest : public CiderArrowFormatJoinTestBase {
```
Comment (jikunshang) [inline cider/tests/functionality/CiderNonEqualLoopJoinTest.cpp:28]: Please keep vanilla test cases here.

File: cider/tests/utils/CiderTestBase.cpp [hunk 3]
```diff
@@ -179,6 +180,14 @@ void CiderJoinTestBase::assertJoinQueryRowEqualForArrowFormat(
...
 
-  EXPECT_TRUE(CiderBatchChecker::checkArrowEq(
-      duck_res_batches[0], cider_res_batch, ignore_order));
+  if (compare_value) {
+    EXPECT_TRUE(CiderBatchChecker::checkArrowEq(
+        duck_res_batches[0], cider_res_batch, ignore_order));
+  } else {
+    // only check row_num, column_num since duckdb in will return wrong results in left
+    // range join with nulls.
+    EXPECT_TRUE(duck_res_batches[0]->getChildrenNum() ==
```
Comment (jikunshang) [inline cider/tests/utils/CiderTestBase.cpp:189]: EXPECT_EQ

File: cider/tests/utils/CiderTestBase.cpp [hunk 4]
```diff
@@ -168,7 +168,8 @@ void CiderJoinTestBase::assertJoinQueryRowEqual(const std::string& sql,
 void CiderJoinTestBase::assertJoinQueryRowEqualForArrowFormat(
     const std::string& sql,
     const std::string& json_file,
-    const bool ignore_order) {
+    const bool ignore_order,
+    const bool compare_value) {
   VLOG(4) << sql;
```
Comment (jikunshang) [inline cider/tests/utils/CiderTestBase.cpp:173]: remove this?

File: cider/tests/functionality/CiderNonEqualLoopJoinTest.cpp (hunk 1 above)
Comment (spevenhe) [reply cider/tests/functionality/CiderNonEqualLoopJoinTest.cpp:27]: Even if we create NOT NULL column, in left join cases, there should be null results when left table failed to find a match in right table. But current columnvar codegen will not go to null handle part if we set column for NOT NULL.

File: cider/tests/utils/CiderTestBase.cpp (hunk 4 above)
Comment (spevenhe) [reply cider/tests/utils/CiderTestBase.cpp:173]: This is debug4 level log to help identify the sql when using ```define``` for test. And in cider codegen part, it also contains many debug level log.

File: cider/exec/template/IRCodegen.cpp [hunk 5]
```diff
@@ -717,26 +717,58 @@ std::vector<JoinLoop> Executor::buildJoinLoops(
...
-            for (auto expr : it->second) {
-              left_join_cond = cgen_state_->ir_builder_.CreateAnd(
-                  left_join_cond,
-                  code_generator.toBool(
-                      code_generator.codegen(expr.get(), true, co).front()));
+    std::function<llvm::Value*(const std::vector<llvm::Value*>&)>
+        outer_join_condition_remaining_quals_cb;
+    if (co.use_cider_data_format) {
+      outer_join_condition_remaining_quals_cb =
+          [this, level_idx, &co](const std::vector<llvm::Value*>& prev_iters) {
```
Comment (harborn) [inline cider/exec/template/IRCodegen.cpp:724]: why not implement this lambda as a normal function?
if the lambda convert to a `std::function<llvm::Value*(const std::vector<llvm::Value*>&)>`
the std::function's execution efficiency is really slow.

Comment (spevenhe) [reply cider/exec/template/IRCodegen.cpp:724]: this part would refactor in JITLib, so does not care about efficiency right now.

File: cider/tests/functionality/CiderNonEqualLoopJoinTest.cpp (hunk 2 above)
Comment (spevenhe) [reply cider/tests/functionality/CiderNonEqualLoopJoinTest.cpp:101]: added CiderArrowSeqNotNullInDDLJoinTest

Comment (spevenhe) [reply cider/tests/functionality/CiderNonEqualLoopJoinTest.cpp:101]: updated. Add original tests

**Review by jikunshang — APPROVED** [review]

**Review by yma11 — APPROVED** [review]

**Pull request merged**